File: app/src/main/java/com/example/groxweatherapp/grox/WeatherErrorAction.java
package com.example.groxweatherapp.grox;

import com.example.groxweatherapp.api.WeatherApiClient.RequestMode;
import com.example.groxweatherapp.model.WeatherModel;
import com.groupon.grox.Action;
import java.util.ArrayList;
import java.util.Collections;

import static com.example.groxweatherapp.model.WeatherModel.WeatherModelState.ERROR;

public class WeatherErrorAction implements Action<WeatherModel> {

    @RequestMode private final int requestMode;

    public WeatherErrorAction(Throwable throwable, @RequestMode int requestMode) {
        this.requestMode = requestMode;
    }

    @Override
    public WeatherModel newState(WeatherModel oldState) {
        return oldState.toBuilder()
            .setState(ERROR)
            .setRequestMode(requestMode)
            .setWeather(Collections.unmodifiableList(new ArrayList<>()))
            .build();
    }
}

File: app/src/main/java/com/example/groxweatherapp/grox/WeatherErrorAction.kt
package com.example.groxweatherapp.grox

import com.example.groxweatherapp.api.WeatherApiClient.Companion.ForecastMode
import com.example.groxweatherapp.model.WeatherModel
import com.example.groxweatherapp.model.WeatherModel.Companion.WeatherModelState.ERROR
import com.groupon.grox.Action


class WeatherErrorAction(@ForecastMode private val forecastMode: Int) : Action<WeatherModel> {

    override fun newState(oldState: WeatherModel): WeatherModel {
        return oldState.copy(
                state = ERROR,
                forecastMode = forecastMode,
                weather = emptyList())
    }
}
